text/0219-ecs-service-extensions.md

```diff
@@ -0,0 +1,503 @@
+modify the `Service` properties, `TaskDefinition` properties, and/or make use of the
+resulting `Service` and `TaskDefinition`. Each `ServiceExtension` can implement the
+following hooks:
+
+- `addHooks()` - This hook is called after all the extensions are added to a
+  `ServiceDescription`, but before any of the other extension hooks have been run. It gives
+  each extension a chance to do some inspection of the overall `ServiceDescription` and see
+  what other extensions have been added. Some extensions may want to register hooks on the
+  other extensions to modify them. For example the Firelens extension wants to be able to
+  modify the settings of the application container to route logs through Firelens.
+- `modifyTaskDefinitionProps()` - This is hook is passed the proposed `ecs.TaskDefinitionProps`
+  for a TaskDefinition that is about to be created. This allows the extension to make
+  modifications to the task definition props before the `TaskDefinition` is created.
+  For example the App Mesh extension modifies the proxy settings for the task.
+- `useTaskDefinition()` - After the `TaskDefinition` is created, this hook is passed the
+  actual `TaskDefinition` construct that was created. This allows the extension to add
+  containers to the task, modify the task definition's IAM role, etc.
+- `resolveContainerDependencies()` - Once all extensions have added their containers each
+  extension is given a chance to modify its container's `dependsOn` settings. Extensions
+  need to check and see what other extensions were enabled, and decide whether their
+  container needs to wait on another container to start first.
+- `modifyServiceProps()` - Before an `Ec2Service` or `FargateService` is created this hook
+  is passed a draft version of the service props to change. Each extension adds its own
+  modifications to the service properties. For example the App Mesh extension needs to
+  modify the service settings to enable CloudMap service discovery.
+- `useService()` - After the service is created this hook is given a chance to utilize
+  the service that was created. This is used by extensions like the load balancer or App
+  Mesh extension which create and link other AWS resources to the ECS extension.
+- `connectToService()` - This hook is called when a user wants to connect one service to
+  another service. It allows an extension to implement logic about how to allow connections
+  from one service to another. For example the App Mesh extension implements this method
+  so that you can easily connect one service mesh service to another so that the service's
+  Envoy proxy sidecars know how to route traffic to each other.
+
+The `ServiceExtension` is a basic abstract class which can be extended like this:
+
+```js
+export class MyCustomExtension extends ServiceExtension {
+  constructor() {
+    super('my-extension');
+  }
+
+  public mutateServiceProps(props: ServiceBuild) {
+    return {
+      ...props,
```

Comment:
This pattern is convenient in TypeScript/JavaScript but will not translate to other languages well.

Something to keep in mind...